core/src/main/java/org/apache/carbondata/core/mutate/CarbonUpdateUtil.java

```diff
@@ -733,9 +733,19 @@ public static long cleanUpDeltaFiles(CarbonTable table, boolean isDryRun) throws
       }
       SegmentUpdateDetails[] updateDetails = updateStatusManager.readLoadMetadata();
       for (SegmentUpdateDetails block : updateDetails) {
-        totalDeltaFiles.stream().filter(fileName -> fileName.getName().endsWith(block
-                .getDeleteDeltaStartTimestamp() + CarbonCommonConstants.DELETE_DELTA_FILE_EXT))
-                .collect(Collectors.toList()).forEach(fileName -> totalDeltaFiles.remove(fileName));
+        // Case 1: When deleteDeltaStartTimestamp = deleteDeltaEndTimestamp. in this case only 1
+        // delta file is present and deltaFileStamps is NULL
+        // Case 2: When deleteDeltaStartTimestamp != deleteDeltaEndTimestamp. in this case more
+        // than 1 delta files are present, then can blindly read deltaFilesStamps variable
+        if (block.getDeleteDeltaStartTimestamp().equals(block.getDeleteDeltaEndTimestamp())) {
+          totalDeltaFiles.stream().filter(fileName -> fileName.getName().endsWith(block
+                  .getDeleteDeltaStartTimestamp() + CarbonCommonConstants.DELETE_DELTA_FILE_EXT))
+                  .collect(Collectors.toList()).forEach(totalDeltaFiles::remove);
+        } else {
+          block.getDeltaFileStamps().stream().forEach(fileName -> totalDeltaFiles
+                  .removeIf(filter -> filter.getName().endsWith(fileName +
+                          CarbonCommonConstants.DELETE_DELTA_FILE_EXT)));
+        }
```

Comment:
instead of streaming for each `block`, can merge and do single looping?

Reply:
done